src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -421,8 +421,12 @@ private static BoundDagTemp OriginalInput(BoundDagTemp input)
             ArrayBuilder<Tests> tests)
         {
             // Add a null test if needed
-            if (input.Type.CanContainNull())
+            if (input.Type.CanContainNull() &&
+                // The slice value is assumed to be never null
+                input.Source is not BoundDagSliceEvaluation)
```

Comment:
Here's the current behavior: it is safe (no null-ref exception thrown because we do emit a null check) but annoying (an arguably incorrect subsumption error). I think that's an acceptable behavior.

```
#nullable enable

if (new C() is [_, ..[0, _]]) // doesn't crash (IL is correct)
{
    
}

_ = new C() switch
{
   [_, 0, _] => 0,
   [_, ..[0, _]] => 0, // error CS8510: The pattern is unreachable.
   _ => 0,
};

class C
{
    public int Length => 3;
    public int this[int i] => 0;
    public C? Slice(int i, int j) => null;
}
```

Reply:
Let's add a test for the scenario you pointed out offline (difference between `Span<int>` and `int[]`) [sharplab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQBsCWyaoA6wwQIB2yAPgAIBMAjALABQAygSQDyYkwB8ABMj4BePgBMIYKHHQwA3E2QB3TDADGACz4AKZAEo+AbyZ8TfVVGQQ+AbQB0t67QC6AGmouAzE5DHT5yza0bp7efAD0YXzECAD2CL6mpsBIUADW8owAvkxMXDDWTnxQwmISUjIZSioa2lD6RoyJ/lZ2Ds7BXj6NfhYtQe6d4ZEkMVEIsfHdiSbJ0OlM2YxAA==)